engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,121 @@
+    val messageFormatter = new ConfluentAvroMessageFormatter(schemaRegistryClient.client)
+    val createReader = (topic: String) => new ConfluentAvroMessageReader(schemaRegistryClient.client, topic)
+
+    new ConfluentAvroToJsonFormatter(kafkaConfig, messageFormatter, createReader, kafkaSourceDeserializationSchema)
+  }
+
+}
+
+/**
+  * @tparam K - key type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+  * @tparam V - value type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+  */
+class ConfluentAvroToJsonFormatter[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig,
+                                                             messageFormatter: ConfluentAvroMessageFormatter,
+                                                             createReader: String => ConfluentAvroMessageReader,
+                                                             deserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]]
+                                                            ) extends RecordFormatter {
+
+  /**
+    * Step 1: Deserialize raw kafka event to GenericRecord/SpecificRecord domain.
+    * Step 2: Create Encoders that use ConfluentAvroMessageFormatter to convert avro object to json
+    * Step 3: Encode event's data with schema id's with derived encoder.
+    */
+  override protected def formatRecord(record: ConsumerRecord[Array[Byte], Array[Byte]]): Array[Byte] = {
+    val deserializedRecord = deserializationSchema.deserialize(record)
+    val serializableRecord = AvroSerializableConsumerRecord(
+      messageFormatter.getSchemaIdOpt(record.key()),
+      messageFormatter.getSchemaIdOpt(record.value()).get,
+      SerializableConsumerRecord(deserializedRecord)
+    )
+
+    consumerRecordEncoder(serializableRecord).noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  /**
+    * Step 1: Deserialize raw json bytes to AvroSerializableConsumerRecord[Json, Json] domain without interpreting key and value content.
+    * Step 2: Create key and value json-to-avro interpreter based on schema id's provided in json.
+    * Step 3: Use interpreter to create raw kafka ConsumerRecord
+    */
+  override protected def parseRecord(topic: String, bytes: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    val record = CirceUtil.decodeJsonUnsafe(bytes)(consumerRecordDecoder)
+
+    def serializeKeyValue(keyOpt: Option[Json], value: Json): (Array[Byte], Array[Byte]) = {
+      val reader = createReader(topic)
+      val keyBytes = if (kafkaConfig.useStringForKey) {
+        keyOpt match {
+          // we handle strings this way because we want to keep result value compact and JString is formatted in quotes
+          case Some(j) if j.isString => j.asString.get.getBytes(StandardCharsets.UTF_8)
+          case Some(j) => j.noSpaces.getBytes(StandardCharsets.UTF_8)
```

Comment:
Why this case must be handled? Don't we always have raw "String" for this flag?

Reply:
Fixed.